File: src/main/java/com/github/xumeng/java/problems/easy/problem0257/Solution0257.java
package com.github.xumeng.java.problems.easy.problem0257;

import com.github.xumeng.java.common.datastructure.TreeNode;

import java.util.ArrayList;
import java.util.List;

public class Solution0257 {

    public List<String> binaryTreePaths(TreeNode root) {
        List<String> paths = new ArrayList<>();
        if (root == null) {
            return paths;
        }
        if (root.left == null && root.right == null) {
            paths.add(String.valueOf(root.val));
        }
        if (root.left != null) {
            List<String> pathsOfLeft = binaryTreePaths(root.left);
            for (String path : pathsOfLeft) {
                paths.add(root.val + "->" + path);
            }
        }
        if (root.right != null) {
            List<String> pathsOfRight = binaryTreePaths(root.right);
            for (String path : pathsOfRight) {
                paths.add(root.val + "->" + path);
            }
        }
        return paths;
    }

}

File: src/main/kotlin/com/github/xumeng/kotlin/problems/easy/problem0257/Solution0257.kt
package com.github.xumeng.kotlin.problems.easy.problem0257

import com.github.xumeng.kotlin.common.datastructure.TreeNode

class Solution0257 {

    fun binaryTreePaths(root: TreeNode?): List<String> {
        val paths: MutableList<String> = mutableListOf()
        if (root == null) {
            return paths
        }
        if (root.left == null && root.right == null) {
            paths.add(root.value.toString())
        }
        if (root.left != null) {
            val pathsOfLeft: List<String> = binaryTreePaths(root.left)
            for (path: String in pathsOfLeft) {
                paths.add(root.value.toString() + "->" + path)
            }
        }
        if (root.right != null) {
            val pathsOfRight: List<String> = binaryTreePaths(root.right)
            for (path: String in pathsOfRight) {
                paths.add(root.value.toString() + "->" + path)
            }
        }
        return paths
    }

}